docs/settings.rst
.. _ref-settings:

=================
Haystack Settings
=================

As a way to extend/change the default behavior within Haystack, there are
several settings you can alter within your ``settings.py``. This is a
comprehensive list of the settings Haystack recognizes.


``HAYSTACK_DEFAULT_OPERATOR``
=============================

**Optional**

This setting controls what the default behavior for chaining ``SearchQuerySet``
filters together is.

Valid options are::

    HAYSTACK_DEFAULT_OPERATOR = 'AND'
    HAYSTACK_DEFAULT_OPERATOR = 'OR'

Defaults to ``AND``.


``HAYSTACK_CONNECTIONS``
========================

**Required**

This setting controls which backends should be available. It should be a
dictionary of dictionaries resembling the following (complete) example::

    HAYSTACK_CONNECTIONS = {
        'default': {
            'ENGINE': 'haystack.backends.solr_backend.SolrEngine',
            'URL': 'http://localhost:8983/solr/default',
            'TIMEOUT': 60 * 5,
            'INCLUDE_SPELLING': True,
            'BATCH_SIZE': 100,
            'EXCLUDED_INDEXES': ['thirdpartyapp.search_indexes.BarIndex'],
        },
        'autocomplete': {
            'ENGINE': 'haystack.backends.whoosh_backend.WhooshEngine',
            'PATH': '/home/search/whoosh_index',
            'STORAGE': 'file',
            'POST_LIMIT': 128 * 1024 * 1024,
            'INCLUDE_SPELLING': True,
            'BATCH_SIZE': 100,
            'EXCLUDED_INDEXES': ['thirdpartyapp.search_indexes.BarIndex'],
        },
        'slave': {
            'ENGINE': 'xapian_backend.XapianEngine',
            'PATH': '/home/search/xapian_index',
            'INCLUDE_SPELLING': True,
            'BATCH_SIZE': 100,
            'EXCLUDED_INDEXES': ['thirdpartyapp.search_indexes.BarIndex'],
        },
        'db': {
            'ENGINE': 'haystack.backends.simple_backend.SimpleEngine',
            'EXCLUDED_INDEXES': ['thirdpartyapp.search_indexes.BarIndex'],
        }
    }

No default for this setting is provided.

The main keys (``default`` & friends) are identifiers for your application.
You can use them any place the API exposes ``using`` as a method or kwarg.

There must always be at least a ``default`` key within this setting.

The ``ENGINE`` option is required for all backends & should point to the
``BaseEngine`` subclass for the backend.

Additionally, each backend may have additional options it requires:

* Solr

  * ``URL`` - The URL to the Solr core. e.g. http://localhost:8983/solr/collection1
  * ``ADMIN_URL`` - The URL to the administrative functions. e.g.
    http://localhost:8983/solr/admin/cores

* Whoosh

  * ``PATH`` - The filesystem path to where the index data is located.

* Xapian

  * ``PATH`` - The filesystem path to where the index data is located.

The following options are optional:

* ``INCLUDE_SPELLING`` - Include spelling suggestions. Default is ``False``
* ``BATCH_SIZE`` - How many records should be updated at once via the management
  commands. Default is ``1000``.
* ``TIMEOUT`` - (Solr and ElasticSearch) How long to wait (in seconds) before
  the connection times out. Default is ``10``.
* ``STORAGE`` - (Whoosh-only) Which storage engine to use. Accepts ``file`` or
  ``ram``. Default is ``file``.
* ``POST_LIMIT`` - (Whoosh-only) How large the file sizes can be. Default is
  ``128 * 1024 * 1024``.
* ``FLAGS`` - (Xapian-only) A list of flags to use when querying the index.
* ``EXCLUDED_INDEXES`` - A list of strings (as Python import paths) to indexes
  you do **NOT** want included. Useful for omitting third-party things you
  don't want indexed or for when you want to replace an index.
* ``KWARGS`` - (Solr and ElasticSearch) Any additional keyword arguments that
  should be passed on to the underlying client library.
* ``DATE_FACET_FIELD`` - (Solr-only) Support to ``date_facet`` on Solr >= 6.6.
  Olders set ``date``. Default is ``range``.


``HAYSTACK_ROUTERS``
====================

**Optional**

This setting controls how routing is performed to allow different backends to
handle updates/deletes/reads.

An example::

    HAYSTACK_ROUTERS = ['search_routers.MasterSlaveRouter', 'haystack.routers.DefaultRouter']

Defaults to ``['haystack.routers.DefaultRouter']``.


``HAYSTACK_SIGNAL_PROCESSOR``
=============================

**Optional**

This setting controls what ``SignalProcessor`` class is used to handle Django's
signals & keep the search index up-to-date.

An example::

    HAYSTACK_SIGNAL_PROCESSOR = 'haystack.signals.RealtimeSignalProcessor'

Defaults to ``'haystack.signals.BaseSignalProcessor'``.


``HAYSTACK_DOCUMENT_FIELD``
===========================

**Optional**

This setting controls what fieldname Haystack relies on as the default field
for searching within.

An example::

    HAYSTACK_DOCUMENT_FIELD = 'wall_o_text'

Defaults to ``text``.


``HAYSTACK_SEARCH_RESULTS_PER_PAGE``
====================================

**Optional**

This setting controls how many results are shown per page when using the
included ``SearchView`` and its subclasses.

An example::

    HAYSTACK_SEARCH_RESULTS_PER_PAGE = 50

Defaults to ``20``.


``HAYSTACK_CUSTOM_HIGHLIGHTER``
===============================

**Optional**

This setting allows you to specify your own custom ``Highlighter``
implementation for use with the ``{% highlight %}`` template tag. It should be
the full path to the class.

An example::

    HAYSTACK_CUSTOM_HIGHLIGHTER = 'myapp.utils.BorkHighlighter'

No default is provided. Haystack automatically falls back to the default
implementation.


``HAYSTACK_ITERATOR_LOAD_PER_QUERY``
====================================

**Optional**

This setting controls the number of results that are pulled at once when
iterating through a ``SearchQuerySet``. If you generally consume large portions
at a time, you can bump this up for better performance.

.. note::

    This is not used in the case of a slice on a ``SearchQuerySet``, which
    already overrides the number of results pulled at once.

An example::

    HAYSTACK_ITERATOR_LOAD_PER_QUERY = 100

The default is 10 results at a time.


``HAYSTACK_LIMIT_TO_REGISTERED_MODELS``
=======================================

**Optional**

This setting allows you to control whether or not Haystack will limit the
search results seen to just the models registered. It should be a boolean.

If your search index is never used for anything other than the models
registered with Haystack, you can turn this off and get a small to moderate
performance boost.

An example::

    HAYSTACK_LIMIT_TO_REGISTERED_MODELS = False

Default is ``True``.


``HAYSTACK_ID_FIELD``
=====================

**Optional**

This setting allows you to control what the unique field name used internally
by Haystack is called. Rarely needed unless your field names collide with
Haystack's defaults.

An example::

    HAYSTACK_ID_FIELD = 'my_id'

Default is ``id``.


``HAYSTACK_DJANGO_CT_FIELD``
============================

**Optional**

This setting allows you to control what the content type field name used
internally by Haystack is called. Rarely needed unless your field names
collide with Haystack's defaults.

An example::

    HAYSTACK_DJANGO_CT_FIELD = 'my_django_ct'

Default is ``django_ct``.


``HAYSTACK_DJANGO_ID_FIELD``
============================

**Optional**

This setting allows you to control what the primary key field name used
internally by Haystack is called. Rarely needed unless your field names
collide with Haystack's defaults.

An example::

    HAYSTACK_DJANGO_ID_FIELD = 'my_django_id'

Default is ``django_id``.


``HAYSTACK_IDENTIFIER_METHOD``
==============================

**Optional**

This setting allows you to provide a custom method for
``haystack.utils.get_identifier``. Useful when the default identifier
pattern of <app.label>.<object_name>.<pk> isn't suited to your
needs.

An example::

    HAYSTACK_IDENTIFIER_METHOD = 'my_app.module.get_identifier'

Default is ``haystack.utils.default_get_identifier``.


``HAYSTACK_FUZZY_MIN_SIM``
==========================

**Optional**

This setting allows you to change the required similarity when using ``fuzzy``
filter.

Default is ``0.5``


``HAYSTACK_FUZZY_MAX_EXPANSIONS``
=================================

**Optional**

This setting allows you to change the number of terms fuzzy queries will
expand to when using ``fuzzy`` filter.

Default is ``50``


docs/changelog.rst
Changelog
=========


%%version%% (unreleased)
------------------------
- Docs: don't tell people how to install Python packages. [Chris Adams]

  It's 2018, "pip install <packagename>" is the only thing we should
  volunteer.
- Update Elasticsearch documentation. [Chris Adams]

  * Add 5.x to supported versions
  * Replace configuration and installation information with
    pointers to the official docs
  * Stop mentioning pyes since it’s fallen behind the official
    client in awareness
  * Don’t tell people how to install Python packages
- Fix get_coords() calls. [Chris Adams]
- Update README & contributor guide. [Chris Adams]
- Blacken. [Chris Adams]
- Isort everything. [Chris Adams]
- Update code style settings. [Chris Adams]

  Prep for Blackening
- Remove PyPy / Django 2 targets. [Chris Adams]

  We'll restore these when pypy3 is more mainstream
- Use default JRE rather than requiring Oracle. [Chris Adams]

  OpenJDK is also supported and that does not require accepting a license.
- Changed ES5.x test skip message to match the friendlier 2.x one.
  [Bruno Marques]
- Fixed faceted search and autocomplete test. [Bruno Marques]
- Removed ES5 code that actually never runs. [Bruno Marques]
- Fixed kwargs in ES5's build_search_query. [Bruno Marques]
- ES5: fixed MLT, within and dwithin. [Bruno Marques]
- Assorted ES5.x fixes. [Bruno Marques]
- Re-added sorting, highlighting and suggesting to ES5.x backend. [Bruno
  Marques]
- Fixed filters and fuzziness on ES5.x backend. [Bruno Marques]
- Added Java 8 to Travis dependencies. [Bruno Marques]
- Started Elasticsearch 5.x support. [Bruno Marques]
- Style change to avoid ternary logic on the end of a line. [Chris
  Adams]

  This is unchanged from #1475 but avoids logic at the end of the line
- Do not raise when model cannot be searched. [benvand]

  * Return empty string.
  * Test.
- Merge pull request #1616 from hornn/batch_order. [Chris Adams]

  Order queryset by pk in update batching
- Order queryset by pk in update batching This solves #1615. [Noa Horn]

  The queryset is not ordered by pk by default, however the batching filter relies on the results being ordered.
  When the results are not ordered by pk, some objects are not indexed.
  This can happen when the underlying database doesn't have default ordering by pk, or when the model or index_queryset() have a different ordering.
- Merge pull request #1612 from hornn/patch-1. [Chris Adams]

  Construct django_ct based on model instead of object
- Update indexes.py. [Noa Horn]

  Construct django_ct based on model instead of object.
  This solves issue #1611 - delete stale polymorphic model documents.
- Merge pull request #1610 from erez-o/patch-1. [Chris Adams]

  Update installing_search_engines.rst
- Update installing_search_engines.rst. [Chris Adams]
- Update installing_search_engines.rst. [Erez Oxman]

  Updated docs about Solr 6.X+ "More like this"
- Avoid UnicodeDecodeError when an error occurs while resolving
  attribute lookups. [Chris Adams]

  Thanks to Martin Burchell (@martinburchell) for the patch in #1599
- Fix UnicodeDecodeError in error message. [Martin Burchell]

  Because of the way the default __repr__ works in Django models, we can get a
  UnicodeDecodeError when creating the SearchFieldError if a model does not have
  an attribute. eg:
  UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 11: ordinal not in range(128)
  and this hides the real problem.

  I have left alone the other SearchFieldError in this method because current_obj is always
  None. The error message is a bit strange in this case but it won't suffer from the same problem.
- Add max retries option to rebuild_index, matching update_index. [Chris
  Adams]

  Thanks to @2miksyn for the patch in #1598
- Update rebuild_index.py. [2miksyn]

  Add max-retries argument to rebuild_index managment command. This is useful for debug at development time

  Add Django 2.1 compatibility. [Tim Graham]


v2.8.1 (2018-03-16)
-------------------
- Merge pull request #1596 from klass-ivan/collecting-deep-attr-through-
  m2m. [Chris Adams]

  Fixed collection of deep attributes through m2m relation
- Fixed collection of deep attributes through m2m relation. [Ivan Klass]


v2.8.0 (2018-03-09)
-------------------
- Optimize ElasticSearch backend (closes #1590) [Chris Adams]

  Thanks to @klass-ivan for the patch
- [elasticsearch backend] - Fixed index re-obtaining for every field.
  [Ivan Klass]
- Django 2.0 compatibility (closes #1582) [Chris Adams]

  Thanks to @mpauly and @timgraham for working on this and @dani0805,
  @andrewbenedictwallace, @rabidcicada, @webtweakers, @nadimtuhin, and
  @JonLevischi for testing.
- Implemented TG's review comments. [Martin Pauly]
- Drop support for old django versions. [Martin Pauly]
- For some reason the mock needs to return something. [Martin Pauly]
- Django 2.0 changes to tests. [Martin Pauly]
- Dropped a few unnecessary interactive=False. [Martin Pauly]
- Replace get_coords() by coords in more places. [Martin Pauly]
- Ignore python2 Django2 combination. [Martin Pauly]
- Drop tests for Django < 1.11. [Martin Pauly]
- Update requirements in setup.py. [Martin Pauly]
- Update imports to drop Django 1.8 support. [Martin Pauly]
- Fix intendation error in tox.ini. [Martin Pauly]
- Merge https://github.com/django-haystack/django-haystack. [Martin
  Pauly]
- Added a test for exclusion of M2M fields for ModelSearchIndex. [Martin
  Pauly]
- In Django 2.0 ForeinKeys must have on_delete. [Martin Pauly]
- Assuming that everyone who wants to run these tests upgrades pysolr.
  [Martin Pauly]
- Django 2.0 is not compatible with python 2.7. [Martin Pauly]
- Deal with tuples and strings. [Martin Pauly]
- Fix a bug due to string __version__ of pysolr. [Martin Pauly]
- Fix tox. [Martin Pauly]
- Mocking order. [Martin Pauly]
- Reverse order. [Martin Pauly]
- Update test - the interactive kwarg is only passed to the clear_index
  command. [Martin Pauly]

example_project/settings.py
import os

from django.conf import settings

SECRET_KEY = "CHANGE ME"

# All the normal settings apply. What's included here are the bits you'll have
# to customize.

# Add Haystack to INSTALLED_APPS. You can do this by simply placing in your list.
INSTALLED_APPS = settings.INSTALLED_APPS + ("haystack",)


HAYSTACK_CONNECTIONS = {
    "default": {
        # For Solr:
        "ENGINE": "haystack.backends.solr_backend.SolrEngine",
        "URL": "http://localhost:8983/solr/example",
        "TIMEOUT": 60 * 5,
        "INCLUDE_SPELLING": True,
    },
    "elasticsearch": {
        "ENGINE": "haystack.backends.elasticsearch_backend.ElasticsearchSearchEngine",
        "URL": "http://localhost:9200",
        "INDEX_NAME": "example_project",
    },
    "whoosh": {
        # For Whoosh:
        "ENGINE": "haystack.backends.whoosh_backend.WhooshEngine",
        "PATH": os.path.join(os.path.dirname(__file__), "whoosh_index"),
        "INCLUDE_SPELLING": True,
    },
    "simple": {
        # For Simple:
        "ENGINE": "haystack.backends.simple_backend.SimpleEngine"
    },
    # 'xapian': {
    #     # For Xapian (requires the third-party install):
    #     'ENGINE': 'xapian_backend.XapianEngine',
    #     'PATH': os.path.join(os.path.dirname(__file__), 'xapian_index'),
    # }
}
